File: dfsbot.py

import sys
import select
import time
import random


# current exact position
x = 0.5
y = 0.5
home_x,home_y = 0,0

# last tile "reached" (i.e., being close enough to center)
ty = -1
tx = -1

# set of walls known
walls = set()
for i in range(0,11):
  walls |= {(i,0,i+0,0), (i,11,i+1,11), (0,i,0,i+1), (11,i,11,i+1)}

# DFS tree
plan = []
seen = set()
dead = set()

# introduce ourselves, all friendly like
print("himynameis DFS-bot", flush=True)

# Wait a few seconds for some initial sense data
time.sleep(0.25)

while True:
  # while there is new input on stdin:
  while select.select([sys.stdin,],[],[],0.0)[0]:
    # read and process the next 1-line observation
    obs = sys.stdin.readline()
    obs = obs.split(" ")
    if obs == []: pass
    elif obs[0] == "bot":
      # update our own position
      x = float(obs[1])
      y = float(obs[2])
      # print("comment now at: %s %s" % (x,y), flush=True)
      # update our latest tile reached once we are firmly on the inside of the tile
      if ((int(x) != tx or int(y) != ty) and
          ((x-(int(x)+0.5))**2 + (y-(int(y)+0.5))**2)**0.5 < 0.2):
        tx = int(x)
        ty = int(y)
        if plan == []:
          plan = [(tx,ty)]
          home_x = tx
          home_y = ty
          seen = set(plan)
        #print("comment now at tile: %s %s" % (tx,ty), flush=True)
    elif obs[0] == "wall":
      #print("comment wall: %s %s %s %s" % (obs[1],obs[2],obs[3],obs[4]), flush=True)
      # ensure every wall we see is tracked in our walls set
      x0 = int(float(obs[1]))
      y0 = int(float(obs[2]))
      x1 = int(float(obs[3]))
      y1 = int(float(obs[4]))
      walls |= {(x0,y0,x1,y1)}

  # if we've achieved our goal, update our plan and issue a new command
  if len(plan) > 0 and plan[-1] == (tx,ty):
    seen |= {(tx,ty)} # marking the tile as seen
    # returned back to origin
    if(plan[-1][0] == home_x) and (plan[-1][1] == home_y):
      dead -= {(10,10)}
      seen = {(tx,ty)}

    # if we've hit our opposing corner:
    if(plan[-1][0] == 10) and (plan[-1][1] == 10):
      # mark all other tiles dead, this is our final path, backtrack
      planset = set(plan)
      dead = set()
      for i in range(11):
        for j in range(11):
          if (i,j) not in planset:
            dead |= {(i,j)} # also means, there is only one way to move
      seen = set()

    # if pathing, search through lower, right, top, left children, in that order
    # assumes sufficient sense data, but that may not strictly be true  
    if len(seen) > 0 and (tx,ty+1) not in dead|seen and (tx,ty+1,tx+1,ty+1) not in walls:
      plan.append((tx,ty+1))  # move down
    elif len(seen) > 0 and (tx+1,ty) not in dead|seen and (tx+1,ty,tx+1,ty+1) not in walls:
      plan.append((tx+1,ty))  # move right
    elif len(seen) > 0 and (tx,ty-1) not in dead|seen and (tx,ty,tx+1,ty) not in walls:
      plan.append((tx,ty-1))  # move up
    elif len(seen) > 0 and (tx-1,ty) not in dead|seen and (tx,ty,tx,ty+1) not in walls:
      plan.append((tx-1,ty))  # move left
    else:
    # if we cannot advance or are not pathing currently, backtrack
      dead |= {(tx,ty)}
      plan = plan[:-1]
      # backtrack further, in one command, if we're
      #   returning to start AND its in a straight line:
      #while seen == set() and (plan[-1][0] == tx or plan[-1][1] == ty):
      #  plan = plan[:-1]
      
    # issue a command for the latest plan
    print("toward %s %s" % (plan[-1][0]+0.5, plan[-1][1]+0.5), flush=True)
  
  print("", flush=True)
  time.sleep(0.125)
  

File: dfsbot.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Objects;
import java.util.Scanner;
import java.util.Set;

public class dfsbot {

  public static void main(String[] args) throws InterruptedException, NumberFormatException, IOException {
    double x = 0.5;
    double y = 0.5;

    int ty = -1;
    int tx = -1;

    // Walls
    Set<Wall> walls = new HashSet<Wall>();
    for (int i = 0; i < 11; i++) {
      walls.add(new Wall(i, 0, i + 0, 0));
      walls.add(new Wall(i, 11, i + 1, 11));
      walls.add(new Wall(0, i, 0, i + 1));
      walls.add(new Wall(11, i, 11, i + 1));
    }
    
    // DFS Tree
    ArrayList<Point> plan = new ArrayList<Point>();
    Set<Point> seen = new HashSet<Point>();
    Set<Point> dead = new HashSet<Point>();

    // this should flush the output right away.
    System.out.println("himynameis DFSBot");
    // wait a few seconds for the intial sense data
    Thread.sleep(250);

    Scanner sc = new Scanner(System.in);
    while (true) {
        // while there is new input on the stdin
      while (System.in.available() > 0) {
        // read and process the next 1-line observation from the server
        String obs = sc.nextLine();
        ArrayList<String> obsList = new ArrayList<String>();
        for (String s : obs.split(" ")) {
          obsList.add(s);
        }
        if (obsList.size() == 0) {
          continue;
        } else if (obsList.get(0).equals("bot")) {
            // update our position
          x = Float.parseFloat(obsList.get(1));
          y = Float.parseFloat(obsList.get(2));
          // update our latest tile reached once we are firmly on the inside of the tile
          if (((int) x != tx || (int) y != ty)
              && Math.sqrt((Math.pow((x - ((int) x + 0.5)), 2)) + (Math.pow((y - ((int) y + 0.5)), 2))) < 0.2) {
            tx = (int) x;
            ty = (int) y;
            if (plan.size() == 0) {
              plan.add(new Point(tx, ty));
            }
          }
        } else if (obsList.get(0).equals("wall")) {
            // ensure every wall we see is tracked in our walls set
          int x0 = (int) Double.parseDouble(obsList.get(1));
          int y0 = (int) Double.parseDouble(obsList.get(2));
          int x1 = (int) Double.parseDouble(obsList.get(3));
          int y1 = (int) Double.parseDouble(obsList.get(4));
          walls.add(new Wall(x0, y0, x1, y1));
        }
      }

      if (plan.size() > 0 && (plan.get(plan.size() - 1).equals(new Point(tx, ty)))) {
        seen.add(new Point(tx, ty));
        //  assumes sufficient sense data, but that may not strictly be true 
        if (!(seen.contains(new Point(tx, ty + 1)) || dead.contains(new Point(tx, ty + 1)))
            && !(walls.contains(new Wall(tx, ty + 1, tx + 1, ty + 1)))) {
          plan.add(new Point(tx, ty + 1)); // move down
        } else if (!(seen.contains(new Point(tx + 1, ty)) || dead.contains(new Point(tx + 1, ty)))
            && !(walls.contains(new Wall(tx + 1, ty, tx + 1, ty + 1)))) {
          plan.add(new Point(tx + 1, ty)); // move right
        } else if (!(seen.contains(new Point(tx, ty - 1)) || dead.contains(new Point(tx, ty - 1)))
            && !(walls.contains(new Wall(tx, ty, tx + 1, ty)))) {
          plan.add(new Point(tx, ty - 1)); // move up
        } else if (!(seen.contains(new Point(tx - 1, ty)) || dead.contains(new Point(tx - 1, ty)))
            && !(walls.contains(new Wall(tx, ty, tx, ty + 1)))) {
          plan.add(new Point(tx - 1, ty)); // move left
        } else { // backtrack
          dead.add(new Point(tx, ty));
          plan.remove(plan.size() - 1);
        }
        double x_dir = plan.get(plan.size() - 1).tx + 0.5;
        double y_dir = plan.get(plan.size() - 1).ty + 0.5;
        // issue a new command for the new plan
        System.out.println("toward " + x_dir + " " + y_dir);
      }

      System.out.println("");
      Thread.sleep(125);
    }
  }
}

class Wall {
  public int x0;
  public int y0;
  public int x1;
  public int y1;

  public Wall(int x0, int y0, int x1, int y1) {
    this.x0 = x0;
    this.y0 = y0;
    this.x1 = x1;
    this.y1 = y1;
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof Wall)) {
      return false;
    }
    Wall w = (Wall) o;
    return (this.x0 == w.x0 && this.y0 == w.y0 && this.x1 == w.x1 && this.y1 == w.y1);
  }

  @Override
  public int hashCode() {
    return Objects.hash(this.x0, this.y0, this.x1, this.y1);
  }

  public String toString() {
    return "(" + x0 + "," + y0 + ") -> (" + x1 + "," + y1 + ")";
  }
}

class Point {
  public int tx;
  public int ty;

  public Point(int tx, int ty) {
    this.tx = tx;
    this.ty = ty;
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof Point)) {
      return false;
    }
    Point p = (Point) o;
    return (this.tx == p.tx && this.ty == p.ty);
  }

  @Override
  public int hashCode() {
    return Objects.hash(tx, ty);
  }

  public String toString() {
    return "(" + tx + "," + ty + ")";
  }
}
